File: tictactoe-java/src/main/java/tictactoe/Game.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package tictactoe;

import java.util.ArrayList;
import java.util.Collection;

/**
 *
 * @author wmacevoy
 */
public class Game implements Comparable<Game> {

    public static final int ROWS = 5;
    public static final int COLS = 5;
    public static final int WIN_LENGTH = 4;

    private Mark[][] board = new Mark[ROWS][COLS];
    private State state = State.TURN_O;
    private int unplayed = ROWS * COLS;

    public State getState() {
        return state;
    }

    public static void rowOk(int row) {
        if (row < 0 || row >= ROWS) {
            throw new IndexOutOfBoundsException("row (" + (row) + ") must be between 0 and " + (ROWS - 1));
        }
    }

    public static void colOk(int col) {
        if (col < 0 || col >= COLS) {
            throw new IndexOutOfBoundsException("col (" + (col) + ") must be between 0 and " + (COLS - 1));
        }

    }

    public void moveOk(int row, int col, Mark mark) {
        rowOk(row);
        colOk(col);
        if (board[row][col] == Mark.UNMARKED && state.turn().equals(mark)) {
            return;
        }
        throw new IllegalStateException("move (" + (new Move(mark, row, col)) + ") invalid in current state");
    }

    public void move(int row, int col, Mark mark) {
        moveOk(row, col, mark);
        board[row][col] = mark;
        --unplayed;
        if (win(row, col)) {
            state = mark.win();
        } else if (draw()) {
            state = State.DRAW;
        } else {
            state = mark.other().turn();
        }
    }

    public void unmove(int row, int col) {
        rowOk(row);
        colOk(col);
        switch (board[row][col]) {
            case X:
                ++unplayed;
                board[row][col] = Mark.UNMARKED;
                state = State.TURN_X;
                return;
            case O:
                ++unplayed;
                board[row][col] = Mark.UNMARKED;
                state = State.TURN_O;
                return;
        }
        throw new IllegalStateException();
    }

    public void reset() {
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                board[row][col] = Mark.UNMARKED;
            }
        }
        state = State.TURN_O;
        unplayed = ROWS * COLS;
    }

    public ArrayList<Move> getMoves() {
        ArrayList<Move> ans = new ArrayList<Move>();

        if (state.over()) {
            return ans;
        }
        Mark mark = state.turn();
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                if (board[row][col] == Mark.UNMARKED) {
                    ans.add(new Move(mark, row, col));
                }
            }
        }
        return ans;
    }

    int repeats(int row, int col, int rowDir, int colDir) {
        Mark mark = board[row][col];
        int count = 0;
        while ((row >= 0 && row < ROWS && col >= 0)
                && (col < COLS && board[row][col] == mark)) {
            row += rowDir;
            col += colDir;
            ++count;
        }
        return count;
    }

    int length(int row, int col, int rowDir, int colDir) {
        return repeats(row, col, rowDir, colDir) + repeats(row, col, -rowDir, -colDir) - 1;
    }

    boolean winRow(int row, int col) {
        return length(row, col, 1, 0) >= WIN_LENGTH;
    }

    boolean winCol(int row, int col) {
        return length(row, col, 0, 1) >= WIN_LENGTH;
    }

    boolean winMainDiag(int row, int col) {
        return length(row, col, 1, 1) >= WIN_LENGTH;
    }

    boolean winOffDiag(int row, int col) {
        return length(row, col, 1, -1) >= WIN_LENGTH;
    }

    public boolean win(int row, int col) {
        return winRow(row, col) || winCol(row, col)
                || winMainDiag(row, col) || winOffDiag(row, col);
    }

    public boolean draw() {
        return unplayed == 0;
    }

    public Mark[][] getBoard() {
        Mark[][] ans = new Mark[ROWS][COLS];
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                ans[row][col] = board[row][col];
            }
        }
        return ans;
    }

    public Game() {
        reset();
    }

    public void play(String moves) {
        for (String move : moves.split(" ")) {
            Move.parse(move).play(this);
        }
    }

    @Override
    public String toString() {
        StringBuilder ans = new StringBuilder();

        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                ans.append(board[row][col].toString());
            }
            ans.append("\n");
        }
        return ans.toString();
    }

    public void copyTo(Game target) {
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                target.board[row][col] = board[row][col];
            }
        }
        target.state = state;
        target.unplayed = unplayed;
    }

    public Game clone() {
        Game ans = new Game();
        copyTo(ans);
        return ans;
    }

    @Override
    public int compareTo(Game to) {
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                int a = board[row][col].toInt();
                int b = to.board[row][col].toInt();
                if (a != b) {
                    return a < b ? -1 : 1;
                }
            }
        }
        return 0;
    }

    @Override
    public boolean equals(Object to) {
        return compareTo((Game) to) == 0;
    }

    Game flipRows() {
        Game ans = clone();
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                ans.board[ROWS - row - 1][col] = board[row][col];
            }
        }
        return ans;
    }

    Game flipCols() {
        Game ans = clone();
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                ans.board[row][COLS - col - 1] = board[row][col];
            }
        }
        return ans;
    }

    Game flipRowsAndCols() {
        Game ans = clone();
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                ans.board[ROWS - row - 1][COLS - col - 1] = board[row][col];
            }
        }
        return ans;
    }

    Game transpose() {
        if (ROWS != COLS) {
            return null;
        }
        Game ans = clone();
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                ans.board[col][row] = board[row][col];
            }
        }
        return ans;
    }

    Game offDiagonalTranspose() {
        if (ROWS != COLS) {
            return null;
        }
        Game ans = clone();
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                ans.board[col][row] = board[ROWS - row - 1][COLS - col - 1];
            }
        }
        return ans;
    }

    public Collection<Game> getEquivClass() {
        ArrayList<Game> ans = new ArrayList<Game>();
        ans.add(clone());
        ans.add(flipRows());
        ans.add(flipCols());
        ans.add(flipRowsAndCols());
        if (ROWS == COLS) {
            ans.add(transpose());
            ans.add(offDiagonalTranspose());
        }
        return ans;
    }

}

File: tictactoe/Game.py
from Const import Const
from Move import Move

class Game:
    def over(self):
        return \
            self._state == Const.STATE_WIN_O or \
            self._state == Const.STATE_WIN_X or \
            self._state == Const.STATE_DRAW

    def moveOk(self,row,col,mark):
        Const.rowOk(row)
        Const.colOk(col)
        Const.markOk(mark)
        if (self._state == Const.STATE_TURN_O) and (mark == Const.MARK_O) and \
           self._board[row][col] == Const.MARK_NONE:
           return
        if (self._state == Const.STATE_TURN_X) and (mark == Const.MARK_X) and \
           self._board[row][col] == Const.MARK_NONE:
           return
        raise ValueError("move (" + str(Move(mark,row,col)) + ") invalid in current state")

    def getMoves(self):
        mark = None
        if self._state == Const.STATE_TURN_O:
            mark = Const.MARK_O
        elif self._state == Const.STATE_TURN_X:
            mark = Const.MARK_X

        if mark == None:
            return []

        moves = []
        for row in range(Const.ROWS):
            for col in range(Const.COLS):
                if self._board[row][col] == Const.MARK_NONE:
                    moves.append(Move(row,col,mark))

        return moves

    def _repeats(self,row,col,rowDir,colDir):
        mark = self._board[row][col]        
        count = 0
        while (row >= 0 and row < Const.ROWS and col >= 0) and \
              col < Const.COLS and self._board[row][col]==mark:
            row=row+rowDir
            col=col+colDir
            count = count + 1
        return count

    def _length(self,row,col,rowDir,colDir):
        return self._repeats(row,col,rowDir,colDir)+self._repeats(row,col,-rowDir,-colDir)-1

    def _winRow(self,row,col):
        return self._length(row,col,1,0) >= Const.WIN_LENGTH
    def _winCol(self,row,col):
        return self._length(row,col,0,1) >= Const.WIN_LENGTH
    def _winMainDiag(self,row,col):
        return self._length(row,col,1,1) >= Const.WIN_LENGTH
    def _winOffDiag(self,row,col):
        return self._length(row,col,1,-1) >= Const.WIN_LENGTH

    def _win(self,row,col):
        return self._winRow(row,col) or self._winCol(row,col) or \
               self._winMainDiag(row,col) or self._winOffDiag(row,col)

    def _draw(self):
        return self._unplayed == 0
        
    def move(self,row,col,mark):
        self.moveOk(row,col,mark)
        self._board[row][col]=mark
        self._unplayed = self._unplayed - 1
        if self._win(row,col):
            if mark == Const.MARK_O:
                self._state = Const.STATE_WIN_O
            if mark == Const.MARK_X:
                self._state = Const.STATE_WIN_X
        elif self._draw():
            self._state = Const.STATE_DRAW
        else:
            if mark == Const.MARK_O:
                self._state = Const.STATE_TURN_X
            if mark == Const.MARK_X:
                self._state = Const.STATE_TURN_O

    def unmove(self,row,col):
        Const.rowOk(row)
        Const.colOk(col)
        if self._board[row][col] == Const.MARK_X:
            self._unplayed = self._unplayed + 1
            self._board[row][col] = Const.MARK_NONE
            self._state = Const.STATE_TURN_X
        elif self._board[row][col] == Const.MARK_O:
            self._unplayed = self._unplayed + 1
            self._board[row][col] = Const.MARK_NONE
            self._state = Const.STATE_TURN_O
        else:
            raise ValueError("unmove (" + str(row) + "," + str(col) + ") invalid in current state")

    def reset(self):
        self._board = [[Const.MARK_NONE for col in range(Const.COLS)] for row in range(Const.ROWS)]

        self._state = Const.STATE_TURN_O
        self._unplayed = Const.ROWS*Const.COLS

    def getBoard(self):
        return [[self._board[row][col]  for col in range(Const.COLS)] for row in range(Const.ROWS)]
        
    def getState(self):
        return self._state
    
    def __init__(self):
        self.reset()

    def play(self,moves):
        for move in moves.split():
            Move.parse(move).play(self)

    def __str__(self):
        ans = "\n"
        for row in range(Const.ROWS):
            s="";
            for col in range(Const.COLS):
                s=s+Const.markStr(self._board[row][col])
            ans = ans + s + "\n"
        return ans

    def copyTo(self,target):
        target._board = [[self._board[row][col] for col in range(Const.COLS)] for row in range(Const.ROWS)]
        target._state = self._state
        target._unplayed = self._unplayed

    def clone(self):
        ans = Game()
        self.copyTo(ans)
        return ans

    def __hash__(self):
        tuple = (self._board[k % Const.ROWS][k // Const.ROWS] for k in range(Const.ROWS*Const.COLS))
        return hash(tuple)

    def __eq__(self, other):
        return other != None and (self._board == other._board)

    def __ne__(self, other):
        return not self.__eq__(other)

    def getIndex(self):
        i = 0
        for row in range(Const.ROWS):
            for col in range(Const.COLS):
                board=self._board[row][col]
                x = 0
                if board == Const.MARK_X:
                    x = 1
                elif board == Const.MARK_O:
                    x = 2
                i = 3*i + x
        return i

    def flipRows(self):
        ans=self.clone()
        for row in range(Const.ROWS):
            for col in range(Const.COLS):
                ans._board[Const.ROWS-row-1][col]=self._board[row][col]
        return ans

    def flipCols(self):
        ans = self.clone()
        for row in range(Const.ROWS):
            for col in range(Const.COLS):
                ans._board[row][Const.COLS-col-1]=self._board[row][col]
        return ans

    def flipRowsAndCols(self):
        ans = self.clone()
        for row in range(Const.ROWS):
            for col in range(Const.COLS):
                ans._board[Const.ROWS-row-1][Const.COLS-col-1]=self._board[row][col]
        return ans

    def transpose(self):
        if Const.ROWS != Const.COLS: return None
        ans = self.clone()
        for row in range(Const.ROWS):
            for col in range(Const.COLS):
                ans._board[col][row]=self._board[row][col]
        return ans

    def offDiagonalTranspose(self):
        if Const.ROWS != Const.COLS: return None        
        return self.flipRows().transpose()
    
    def getEquivClass(self):
        return [self.clone(),self.flipRows(),self.flipCols(),self.flipRowsAndCols()]
        
        
